### ISSUE TITLE
9

### ISSUE THREAD

**Issue by bryanjohnabraham** [bryanjohnabraham/ml#11]
# ============================================================
# ADULT INCOME CLASSIFICATION
# Logistic Regression vs Decision Tree
# ============================================================

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

from sklearn.model_selection import train_test_split
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.pipeline import Pipeline

from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier, plot_tree

from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    classification_report,
    confusion_matrix
)

# ============================================================
# 1. LOAD ADULT INCOME DATASET
# ============================================================

url = "https://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.data"

columns = [
    "age",
    "workclass",
    "fnlwgt",
    "education",
    "education-num",
    "marital-status",
    "occupation",
    "relationship",
    "race",
    "sex",
    "capital-gain",
    "capital-loss",
    "hours-per-week",
    "native-country",
    "income"
]

df = pd.read_csv(
    url,
    names=columns,
    na_values="?",
    skipinitialspace=True
)

print("Dataset shape:", df.shape)

print("\nFirst 5 rows:")
print(df.head())

# ============================================================
# 2. PREPROCESSING
# ============================================================

# Remove rows containing missing values
df = df.dropna()

print("\nShape after removing missing values:", df.shape)

# Remove unnecessary whitespace from categorical columns
categorical_columns = [
    "workclass",
    "education",
    "marital-status",
    "occupation",
    "relationship",
    "race",
    "sex",
    "native-country"
]

for col in categorical_columns:
    df[col] = df[col].str.strip()

# ============================================================
# 3. ENCODE TARGET
# ============================================================

# <=50K  -> 0
# >50K   -> 1

df["income"] = df["income"].str.strip()

df["income"] = df["income"].map({
    "<=50K": 0,
    ">50K": 1
})

print("\nIncome distribution:")
print(df["income"].value_counts())

# ============================================================
# 4. SEPARATE FEATURES AND TARGET
# ============================================================

X = df.drop("income", axis=1)
y = df["income"]

# ============================================================
# 5. IDENTIFY NUMERICAL AND CATEGORICAL FEATURES
# ============================================================

numerical_columns = [
    "age",
    "fnlwgt",
    "education-num",
    "capital-gain",
    "capital-loss",
    "hours-per-week"
]

# ============================================================
# 6. TRAIN-TEST SPLIT
# ============================================================

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.20,
    random_state=42,
    stratify=y
)

print("\nTraining samples:", len(X_train))
print("Testing samples:", len(X_test))

# ============================================================
# 7. PREPROCESSING PIPELINE
# ============================================================

preprocessor = ColumnTransformer(
    transformers=[
        (
            "num",
            StandardScaler(),
            numerical_columns
        ),
        (
            "cat",
            OneHotEncoder(
                handle_unknown="ignore"
            ),
            categorical_columns
        )
    ]
)

# ============================================================
# 8. LOGISTIC REGRESSION
# ============================================================

logistic_model = Pipeline(
    steps=[
        (
            "preprocessor",
            preprocessor
        ),
        (
            "classifier",
            LogisticRegression(
                max_iter=1000
            )
        )
    ]
)

print("\nTraining Logistic Regression...")

logistic_model.fit(
    X_train,
    y_train
)

logistic_predictions = logistic_model.predict(X_test)

# ============================================================
# 9. DECISION TREE
# ============================================================

tree_model = Pipeline(
    steps=[
        (
            "preprocessor",
            preprocessor
        ),
        (
            "classifier",
            DecisionTreeClassifier(
                criterion="entropy",
                max_depth=5,
                random_state=42
            )
        )
    ]
)

print("Training Decision Tree...")

tree_model.fit(
    X_train,
    y_train
)

tree_predictions = tree_model.predict(X_test)

# ============================================================
# 10. CALCULATE PERFORMANCE METRICS
# ============================================================

def evaluate_model(name, y_true, y_pred):

    accuracy = accuracy_score(
        y_true,
        y_pred
    )

    precision = precision_score(
        y_true,
        y_pred,
        zero_division=0
    )

    recall = recall_score(
        y_true,
        y_pred,
        zero_division=0
    )

    f1 = f1_score(
        y_true,
        y_pred,
        zero_division=0
    )

    print("\n====================================")
    print(name)
    print("====================================")

    print("Accuracy :", round(accuracy, 4))
    print("Precision:", round(precision, 4))
    print("Recall   :", round(recall, 4))
    print("F1-score :", round(f1, 4))

    return accuracy, precision, recall, f1

logistic_results = evaluate_model(
    "Logistic Regression",
    y_test,
    logistic_predictions
)

tree_results = evaluate_model(
    "Decision Tree",
    y_test,
    tree_predictions
)

# ============================================================
# 11. CLASSIFICATION REPORT
# ============================================================

print("\n\nLOGISTIC REGRESSION CLASSIFICATION REPORT")
print("------------------------------------------")

print(
    classification_report(
        y_test,
        logistic_predictions,
        target_names=[
            "<=50K",
            ">50K"
        ]
    )
)

print("\nDECISION TREE CLASSIFICATION REPORT")
print("-----------------------------------")

print(
    classification_report(
        y_test,
        tree_predictions,
        target_names=[
            "<=50K",
            ">50K"
        ]
    )
)

# ============================================================
# 12. CONFUSION MATRICES
# ============================================================

print("\nLogistic Regression Confusion Matrix:")
print(
    confusion_matrix(
        y_test,
        logistic_predictions
    )
)

print("\nDecision Tree Confusion Matrix:")
print(
    confusion_matrix(
        y_test,
        tree_predictions
    )
)

# ============================================================
# 13. MODEL COMPARISON TABLE
# ============================================================

results = pd.DataFrame({
    "Model": [
        "Logistic Regression",
        "Decision Tree"
    ],

    "Accuracy": [
        logistic_results[0],
        tree_results[0]
    ],

    "Precision": [
        logistic_results[1],
        tree_results[1]
    ],

    "Recall": [
        logistic_results[2],
        tree_results[2]
    ],

    "F1-Score": [
        logistic_results[3],
        tree_results[3]
    ]
})

print("\n\nMODEL COMPARISON")
print("================")

print(
    results.round(4).to_string(index=False)
)

# ============================================================
# 14. PLOT MODEL COMPARISON
# ============================================================

metrics = [
    "Accuracy",
    "Precision",
    "Recall",
    "F1-Score"
]

x = np.arange(len(metrics))
width = 0.35

plt.figure(figsize=(10, 6))

plt.bar(
    x - width/2,
    results.iloc[0][metrics],
    width,
    label="Logistic Regression"
)

plt.bar(
    x + width/2,
    results.iloc[1][metrics],
    width,
    label="Decision Tree"
)

plt.xticks(
    x,
    metrics
)

plt.ylabel("Score")
plt.title("Logistic Regression vs Decision Tree")
plt.ylim(0, 1)

plt.legend()

plt.tight_layout()
plt.show()

# ============================================================
# 15. DECISION TREE VISUALIZATION
# ============================================================

# Get the trained preprocessing object
fitted_preprocessor = tree_model.named_steps["preprocessor"]

# Get the trained decision tree
fitted_tree = tree_model.named_steps["classifier"]

# Get transformed feature names
feature_names = fitted_preprocessor.get_feature_names_out()

plt.figure(figsize=(25, 12))

plot_tree(
    fitted_tree,
    feature_names=feature_names,
    class_names=[
        "<=50K",
        ">50K"
    ],
    filled=True,
    rounded=True,
    fontsize=8
)

plt.title("Decision Tree for Adult Income Prediction")

plt.tight_layout()
plt.show()

# ============================================================
# 16. FEATURE IMPORTANCE OF DECISION TREE
# ============================================================

tree_importance = pd.Series(
    fitted_tree.feature_importances_,
    index=feature_names
)

tree_importance = tree_importance.sort_values(
    ascending=False
)

print("\n\nTOP DECISION TREE FEATURES")
print("==========================")

print(
    tree_importance.head(15)
)

# ============================================================
# 17. PLOT TOP 15 FEATURE IMPORTANCES
# ============================================================

top_features = tree_importance.head(15).sort_values()

plt.figure(figsize=(10, 7))

top_features.plot(
    kind="barh"
)

plt.xlabel("Importance")
plt.ylabel("Feature")
plt.title("Top 15 Decision Tree Feature Importances")

plt.tight_layout()
plt.show()

# ============================================================
# 18. LOGISTIC REGRESSION FEATURE IMPORTANCE
# ============================================================
#
# Logistic Regression does not have "feature_importances_"
# like a Decision Tree.
#
# Instead, we inspect the absolute value of coefficients.
# Larger absolute coefficient = stronger influence on prediction.
# ============================================================

logistic_preprocessor = logistic_model.named_steps[
    "preprocessor"
]

logistic_classifier = logistic_model.named_steps[
    "classifier"
]

logistic_feature_names = (
    logistic_preprocessor.get_feature_names_out()
)

coefficients = pd.Series(
    logistic_classifier.coef_[0],
    index=logistic_feature_names
)

print("\n\nLOGISTIC REGRESSION COEFFICIENTS")
print("================================")

print(
    coefficients.sort_values(
        key=abs,
        ascending=False
    ).head(15)
)

# ============================================================
# 19. TOP POSITIVE AND NEGATIVE LOGISTIC FEATURES
# ============================================================

print("\nTop features associated with >50K:")
print(
    coefficients.sort_values(
        ascending=False
    ).head(10)
)

print("\nTop features associated with <=50K:")
print(
    coefficients.sort_values(
        ascending=True
    ).head(10)
)
